File: React_Native/React_Native/src/pages/sat_pages/ferramentasSat.js
import React, { Component } from "react";
import { Text, View, Alert,Picker,TextInput,Button,DeviceEventEmitter } from "react-native";
import { SafeAreaView } from "react-native-safe-area-context";

import OperacaoSat from '../../services/operacaoSat';
import RetornoSat from '../../services/retornoSat';

import ValidaCodigo from '../../components/ValidacaoCodigo';

export default class FerramentasSat extends Component{
    constructor(props) {
        super(props);    
        this.state = {
            value:'',
            codigoAtivacao: global.MyVar,
        };
    }

    dialogoSat(messageText){
        Alert.alert("Retorno", messageText);
    }

    ferramentasSat(funcao){
        // Salva  o código de ativação para o usuario não precisar ficar digitando em todas as telas
        global.MyVar = this.state.codigoAtivacao
        this.state.value = funcao;

        // CRIA UMA INSTÂNCIA DO COMPONENTE QUE VALIDA O CÓDIGO
        var validationCodigo = new ValidaCodigo();
        
        if(validationCodigo.getValidation(this.state.codigoAtivacao)){

            //ENVIA PARA OPERAÇÃO SAT ONDE VAI CONECTAR COM O SAT DE ACORDO COM A FUNÇÃO
            var operacaoSat = new OperacaoSat();
            args = {'funcao': funcao, 
                    'random': Math.floor(Math.random() * 9999999).toString(),
                    'codigoAtivar': this.state.codigoAtivacao
            };
            operacaoSat.invocarOperacaoSat(args);
            
            DeviceEventEmitter.once("eventFerramenta", event => { 
                var retorno = event.ferramenta;
                var retornoSat = new RetornoSat(args['funcao'], retorno);
                this.dialogoSat(operacaoSat.formataRetornoSat(retornoSat));
            });
        }else{
            this.dialogoSat("Código de Ativação deve ter entre 8 a 32 caracteres!");
        }
    }
    
    render() {
        return (  
             <SafeAreaView>   
                <View style={{ marginBottom:10,justifyContent:'space-between', marginTop: 30 }}>
                    <Text style={{fontSize:20, fontWeight: 'bold'}}>Código de Ativação Sat</Text>
                    <TextInput
                        style={{height: 50,backgroundColor: 'white', fontSize: 17,width:300, marginStart: 30}}                
                        value={this.state.codigoAtivacao}
                        onChangeText={(text) => this.setState({codigoAtivacao: text})}
                    />
                </View>
                     
                <View style={{marginTop: 15, marginStart: 20, marginEnd: 20}}>
                    <Button
                        color= 'gray'
                        title="bloquear sat"
                        onPress={() => this.ferramentasSat("BloquearSat")}
                    />
                </View>

                <View style={{marginTop: 15, marginStart: 20, marginEnd: 20}}>
                    <Button
                        color= 'gray'
                        title="desbloquear sat"
                        onPress={() => this.ferramentasSat("DesbloquearSat")}
                    />
                </View>

                <View style={{marginTop: 15, marginStart: 20, marginEnd: 20}}>
                    <Button
                        color= 'gray'
                        title="extrair log"
                        onPress={() => this.ferramentasSat("ExtrairLog")}
                    />
                </View>

                <View style={{marginTop: 15, marginStart: 20, marginEnd: 20}}>
                    <Button
                        color= 'gray'
                        title="atualizar software"
                        onPress={() => this.ferramentasSat("AtualizarSoftware")}
                    />
                </View>

                <View style={{marginTop: 15, marginStart: 20, marginEnd: 20}}>
                    <Button
                        color= 'gray'
                        title="verificar versão"
                        onPress={() => this.ferramentasSat("Versao")}
                    />
                </View>
            </SafeAreaView>
        )
    }
}

File: Flutter/Flutter - antigo/lib/pages/sat_pages/ferramentasSat.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:flutter_gertec/services/globalValues.dart';
import 'package:flutter_gertec/services/operacaoSat.dart';
import 'package:flutter_gertec/services/retornoSat.dart';
import 'package:flutter_gertec/util/common_code.dart';
import 'package:flutter_gertec/widgets/widgetsgertec.dart';

class PageFerramentaSat extends StatefulWidget {
  @override
  _PageFerramentaSatState createState() => _PageFerramentaSatState();
}

class _PageFerramentaSatState extends State<PageFerramentaSat> {
  // Inicializa o código de ativação com um valor global, para o usuario não precisar ficar digitando
  final codigoAtivacao = TextEditingController(text: GlobalValues.codAtivarSat); // Codigo de ativação do Sat

  CommonGertec commonGertec = new CommonGertec(); //* Classe que possui partes de código comuns em diversas telas

  // Função para validar os valores digitos pelo usuario e realizar a ativação do SAT
  void ferramentasSat(String funcao) async {
    GlobalValues.codAtivarSat = codigoAtivacao.text; // Salva  o código de ativação para o usuario não precisar ficar digitando em todas as telas
    if (funcao == "Versao" || commonGertec.isCodigoValido(codigoAtivacao.text)) {
      //* Chama a função Invocar Operação Sat, que recebe como parâmetro a "operação invocada" e um "Map com as chaves e seus respectivos valores".
      RetornoSat retornoSat = await OperacaoSat.invocarOperacaoSat(
          // Passa como parâmetro um Map de argumentos de valores que deseja enviar
          args: {'funcao': funcao, 'random': Random().nextInt(999999), 'codigoAtivar': codigoAtivacao.text.toString()});

      //? Caso deseje visualizar o retorno completo basta chamar [retornoSat.getResultadoCompleto]

      // O retorno da operação Versão não é tratado, pois retorna somente uma string
      if (funcao == "Versao") {
        WidgetsGertec.dialogoSat(retornoSat.getResultadoCompleto, context: context);
      } else {
        //* Está função [OperacaoSat.formataRetornoSat] recebe como parâmetro a operação realizada e um objeto do tipo RetornoSat
        //* Retorna uma String com os valores obtidos do retorno da Operação já formatados e prontos para serem exibidos na tela
        // Recomenda-se acessar a função e entender como ela funciona
        String retornoFormatado = OperacaoSat.formataRetornoSat(retornoSat: retornoSat);
        WidgetsGertec.dialogoSat(retornoFormatado, context: context);
      }
    } else {
      WidgetsGertec.dialogoSat("Código de Ativação deve ter entre 8 a 32 caracteres!", context: context);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SingleChildScrollView(
        child: Container(
          color: Colors.white,
          padding: EdgeInsets.only(top: 30),
          height: MediaQuery.of(context).size.height,
          width: MediaQuery.of(context).size.width,
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.center,
            mainAxisSize: MainAxisSize.min,
            children: <Widget>[
              SizedBox(
                height: 30,
              ),
              Padding(
                padding: const EdgeInsets.only(left: 10),
                child: Row(
                  mainAxisAlignment: MainAxisAlignment.start,
                  children: <Widget>[
                    Text(
                      "Código de Ativação SAT: ",
                      style: TextStyle(
                        fontSize: 15,
                        fontWeight: FontWeight.bold,
                      ),
                    ),
                    SizedBox(
                      height: 30,
                      width: 140,
                      child: TextFormField(
                        controller: codigoAtivacao,
                        style: TextStyle(
                          fontSize: 14,
                        ),
                      ),
                    ),
                  ],
                ),
              ),
              SizedBox(height: 20),
              WidgetsGertec.buttonStandard("BLOQUEAR SAT", callback: () => ferramentasSat("BloquearSat")),
              WidgetsGertec.buttonStandard("DESBLOQUEAR SAT", callback: () => ferramentasSat("DesbloquearSat")),
              WidgetsGertec.buttonStandard("EXTRAIR LOG", callback: () => ferramentasSat("ExtrairLog")),
              WidgetsGertec.buttonStandard("ATUALIZAR SOFTWARE", callback: () => ferramentasSat("AtualizarSoftware")),
              WidgetsGertec.buttonStandard("VERIFICAR VERSÃO", callback: () => ferramentasSat("Versao")),
            ],
          ),
        ),
      ),
    );
  }
}
